python/cudf/cudf/core/column_accessor.py

```diff
@@ -19,12 +20,60 @@
 
 import cudf
 from cudf.core import column
-from cudf.utils.utils import cached_property, to_flat_dict, to_nested_dict
+from cudf.utils.utils import cached_property
 
 if TYPE_CHECKING:
     from cudf.core.column import ColumnBase
 
 
+class _NestedGetItemDict(dict):
+    """A dictionary whose __getitem__ method accesses nested dicts.
+
+    This class directly subclasses dict for performance, so there are a number
+    of gotchas: 1) the only safe accessor for nested elements is
+    `__getitem__` (all other accessors will fail to perform nested lookups), 2)
+    nested mappings will not exhibit the same behavior (they will be raw
+    dictionaries unless explicitly created to be of this class), and 3) to
+    construct this class you _must_ use `from_zip` to get appropriate treatment
+    of tuple keys.
+    """
+
+    @classmethod
+    def from_zip(cls, data):
+        """Create from zip, specialized factory for nesting."""
+        obj = cls()
+        for key, value in data:
+            d = obj
+            for k in key[:-1]:
+                d = d.setdefault(k, {})
+            d[key[-1]] = value
+        return obj
+
+    def __getitem__(self, key):
+        """Recursively apply dict.__getitem__ to get nested elements of d."""
+        return reduce(dict.__getitem__, key, self)
+
+
+def _to_flat_dict(d):
+    """
+    Convert the given nested dictionary to a flat dictionary
+    with tuple keys.
+    """
+
+    def _inner(d, parents=None):
+        if parents is None:
+            parents = []
+        for k, v in d.items():
+            if not isinstance(v, d.__class__):
+                if parents:
+                    k = tuple(parents + [k])
+                yield (k, v)
+            else:
+                yield from _inner(d=v, parents=parents + [k])
```

Comment:
I think instead of defaulting to `parents=None` and using lists here we can get away with using tuples? I think something like this would work and I believe the tuple functions here should be faster than the list functions as well.
```suggestion
    def _inner(d, parents=tuple()):
        for k, v in d.items():
            if not isinstance(v, d.__class__):
                if parents:
                    k = parents + (k,)
                yield (k, v)
            else:
                yield from _inner(d=v, parents=parents + (k,))
```

Reply:
Nice! Yes, this will work.